michelson.md

```diff
@@ -1050,36 +1066,38 @@ We define `COMPARE` in terms of a `#DoCompare` function.
   rule #DoCompare((Pair A1 _ ), (Pair B1 _ )) => -1                 requires #DoCompare(A1, B1) ==Int -1
   rule #DoCompare((Pair A1 A2), (Pair B1 B2)) => #DoCompare(A2, B2) requires #DoCompare(A1, B1) ==Int 0
   rule #DoCompare((Pair A1 _ ), (Pair B1 _ )) => 1                  requires #DoCompare(A1, B1) ==Int 1
-
-  rule #DoCompare(B1:Bytes, B2:Bytes) => #DoCompare(Bytes2Int(B1, BE, Unsigned), Bytes2Int(B2, BE, Unsigned))
-  rule #DoCompare(#KeyHash(S1), #KeyHash(S2)) => #DoCompare(S1, S2)
-  rule #DoCompare(#Mutez(I1), #Mutez(I2)) => #DoCompare(I1, I2)
-  rule #DoCompare(#Timestamp(I1), #Timestamp(I2)) => #DoCompare(I1, I2)
-  rule #DoCompare(#Address(S1), #Address(S2)) => #DoCompare(S1, S2)
 ```
 
 The `#DoCompare` function requires additional lemmas for symbolic execution.
 
 ```symbolic
+  rule #DoCompare(V, V) => 0 [simplification]
```

Comment:
```suggestion
  rule #DoCompare         (V, V) => 0 [simplification]
  rule #DoCompareBool     (V, V) => 0 [simplification]
  rule #DoCompareInt      (V, V) => 0 [simplification]
  rule #DoCompareStr      (V, V) => 0 [simplification]
  rule #DoCompareAddr     (V, V) => 0 [simplification]
  rule #DoCompareBytes    (V, V) => 0 [simplification]
  rule #DoCompareKeyHash  (V, V) => 0 [simplification]
  rule #DoCompareTimestamp(V, V) => 0 [simplification]
  rule #DoCompareMutez    (V, V) => 0 [simplification]
```

Reply:
The rules for lifting from `#DoCompare` into the other function can get applied first, it seems.

Reply:
i.e. I ended up with this in my path conditions:

```
#And                                                                  
  {                                                                                                            
    false                                                             
  #Equals                                                                                                   
    #DoCompareAddr ( Sender , Sender ) ==Int 0                        
  }   
```

Reply:
Actually, we can (and should) probably avoid this lifting for all but the builtin types, like `Int, String, Bool`. that way the symbolic simplifications further down should be enough. If we insist on lifting Timestamps, Mutezes, Adrresses, and KeyHashes, we have to replicate the lemmas for those too.

Reply:
The reason why I created a separate function for each case was that, for specific sub-cases, the comparison function is _total_. As far as I understand, this allows the backend to be more aggressive in its simplifications.

At the top level, when the function is still partial, I wanted to avoid assuming anything about the inputs except their sorts. Once we descend into a total subfunction, then we can perform all of the usual reasoning that we expect.

Reply:
I have added lemmas for the cases where we don't have this the rule `#DoCompareX(V,V) => 0` already.